### ISSUE TITLE
Using Vue.t() inside a state of Vuex

### ISSUE THREAD

**Issue by bissolli** [kazupon/vue-i18n#101]
Hey guys! I'am trying the following code:

```javascript
import Vue from 'vue'

export default {
    task: {
        status: [
            { id: 'pending', name: Vue.t('pending') },
            { id: 'done', name: 'Done' }
        ]
    }
}
```

That is my state.js which is the state of my VUEX! When I try to use the Vue.t function I have the following error:
> Uncaught TypeError: _vue2.default.t is not a function

### My solution (which I don't think it's the best one also good for performance)

Let my **state.js** like that:
```javascript
import Vue from 'vue'

export default {
    task: {
        status: [
            { id: 'pending', name: 'pending' },
            { id: 'done', name: 'done' }
        ]
    }
}
```

And I've done the following getters (vuex getter)
```javascript
import { map } from 'lodash'
import { trans } from 'utils/helpers/translation' // helper for Vue.t(string)

export const getTaskStatus = ({ task }) => map(task.status, (obj) => {
    return { id: obj.id, name: trans(obj.name) }
})
```

Anyone knows how is the best way to make it work?

**Comment by mygaochunming** [comment]
Hi,may be you can try this { id: 'pending', name: Vue.prototype.t('pending') }

**Comment by gustavobissolli** [comment]
@mygaochunming does not work as well! Same error =/

> Uncaught TypeError: _vue2.default.prototype.t is not a function

**Comment by KingMario** [comment]
You should use `Vue.prototype.$t` not `Vue.prototype.t`

**Comment by gustavobissolli** [comment]
Sorry guys!

About Vue.prototype.$t... does not work as well =/

**Comment by mygaochunming** [comment]
can you show your code where you use the i18n?

**Comment by kazupon** [comment]
Can you provide reproduction code please?

**Comment by tvanier** [comment]
@gustavobissolli is the i18n plugin installed before calling `Vue.t`? Calling `Vue.use(VueI18n)` makes `Vue.t` available. Could the Vuex store/state be loaded before the i18n plugin?

**Comment by rderimay** [comment]
Dont you think, that the best way to add support for i18n to a vuex module would be able to add a locale, like in regular Vue components ??

Being able to do something like this would be awesome !

```
import * as types from '../mutation-types'

// Initial state
const state = {
	files: []
}

// Getters
const getters = {
	files: (state, getters, rootState) => state.files
}

// Actions
const actions = {
}

// Mutations
const mutations = {
	[types.ADD_FILE] (state, { file }) {
		state.files.push(file)
	}
}

// Locales
const locales = {
	en: {
		'upload': 'uplaod',
		'unkown': 'Unknown',
		'uploading': 'Uploading',
		'queued': 'Queued',
		'success': 'Success',
		'error': 'Error'
	},
	fr: {
		'upload': 'Téléchargement',
		'unkown': 'Inconnu',
		'uploading': 'Téléchargement',
		'queued': 'En attente',
		'success': 'Terminé',
		'error': 'Erreur'
	}
}

export default {
	state,
	getters,
	actions,
	mutations,
	locales
}
```

**Comment by kazupon** [comment]
Close (in-activity)

**Comment by tvanier** [comment]
@rderimay I am not sure actual translation strings should be part of the state, maybe just translation keys?
@bissolli could the store contain task status ids only, components would compute status names? This would keep the store independent from the i18n library, and status names would auto-update on language change?

```javascript
computed: {
    taskStatusName() {
        return this.$t(this.$store.state.taskStatusId)
    }
}
```

**Comment by Dsweih** [comment]
Is there any update? I am trying to use the Vue.t in the mutation but not working. How can I do to use vue-i18n in vuex?

**Comment by tvanier** [comment]
As suggested above keeping translation keys (instead of translated strings) in the store and let components call `this.$t` has several advantages.
If you prefer to call `Vue.t` from store mutations, vue-i18n should be plugged before any mutation is committed, have you tried something like this?

```javascript
import Vue from 'vue';
import VueI18n from 'vue-i18n';
import Vuex from 'vuex';

// define Vue.t
Vue.use(VueI18n);

// Vue.t should now be available in store operations
Vue.use(Vuex);

const store = new Vuex.Store({
  mutations: {
    setTranslatedString(state, payload) {
      state.translatedString = Vue.t(payload.translationKey);
    }
  }
});
```

**Comment by Dsweih** [comment]
I have tried the other options as my store is imported from the other file. I always got an error message (`_vue2.default.t is not a function`) when I called `Vue.t` in `/store.js`. The `Vue` instances are not the same in different files, are they?
Thank you @tvanier anyway.

**Comment by Jones-S** [comment]
Still I don't find a solution for the problem:

In my store file `data.js` I want to use a translation string from i18n.
I am currently using the i18n as a plugin:
```
import Vue from 'vue'
import VueI18n from 'vue-i18n'

Vue.use(VueI18n)

export default ({ app, store }) => {
  app.i18n = new VueI18n({
    locale: store.state.currentLang,
    fallbackLocale: 'de',
    messages: {
      'de': require('~/locales/de.json')
    }
  })
}
```

But all the options from above
(importing VueI18n again and then using `Vue.use(VueI18n);` or writing `Vue.t()` or `Vue.prototype.t()` or `Vue.prototype.$t`, it all does not work.
It either says $t is not a function or that i18n is missing.

Does someone have a solid solution for this problem?
Thanks in advance. Cheers

**Comment by laurentpayot** [comment]
@Jones-S maybe this thread will help you: https://github.com/kazupon/vue-i18n/issues/149

**Comment by smakman** [comment]
You can access the current Vue instance in your Vuex store with `this._vm`, so it should be something like `this._vm.$t('pending')`.

**Comment by Jones-S** [comment]
Thank you for the hints. Quite some time has passed since I had to face this problem. I will try this out as soon as I work on a localized project again. 👍

**Comment by kostia7alania** [comment]
prop is isset but don't work -> `this._vm.$t('pending') `

**Comment by constagorgan** [comment]
**TL;DR:** This worked for me: `this.$t('navigationLinkMain')`

In the component's file (e.g.: **Component.vue**):
```
<template>
  <div>{{ links.navigationLinkMain }}</div>
</template>

<script>
export default {
  name: 'Component',
  data() {
    return {
      links: {
        navigationLinkMain: this.$t('navigationLinkMain')
      }
    };
  }
};
</script>
```

In **i18n.js** file:
```
import Vue from 'vue';
VueI18n from 'vue-i18n';

Vue.use(VueI18n);

export const i18n = new VueI18n({
  locale: 'en', // set locale
  fallbackLocale: 'ro', // set fallback locale
  messages: {
    'en': {
      'navigationLinkMain': 'inreg.co.uk',
    },
    'ro': {
      'navigationLinkMain': 'inreg.ro',
    }
  } // set locale messages
});
```
And in **src/main.js**:
```
(...)
import { i18n } from '@/plugins/i18n';
(...)

new Vue({
  (...)
  i18n,
  render: h => h(App)
}).$mount('#app')
```

**Comment by raphaelarias** [comment]
My current workaround is:

`
let $t = null;

if (process.client) {
  window.onNuxtReady(($nuxt) => {
    $t = $nuxt.$t;
  });
}
`
